Remove CMS deprecation warnings from RecoBTag/PerformanceDB

Dr15Jones opened the pull request:
#### PR description:

- changed legacy module types to thread-friendly types
- added esConsumes

#### PR validation:

Code compiles.

Dr15Jones committed:
Remove CMS deprecation warnings from RecoBTag/PerformanceDB

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-36576/27531

- This PR adds an extra 24KB to repository

cmsbuild commented:
A new Pull Request was created by  @Dr15Jones (Chris Jones) for master.

It involves the following packages:

- RecoBTag/PerformanceDB (**reconstruction**)


@jpata, @cmsbuild, @clacaputo, @slava77 can you please review it and eventually sign? Thanks.
@JyothsnaKomaragiri, @emilbols, @demuller, @andrzejnovak, @AlexDeMoor this is something you requested to watch as well.
@perrotta, @dpiparo, @qliphy you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

Dr15Jones commented:
please test

Dr15Jones commented:
NOTE: this package appears to exist solely to deliver the EventSetup data product `BtagPerformance` which is not used anywhere else in CMSSW. It appears to me like this whole package could be removed.

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-2ab928/21442/summary.html
**COMMIT**: 62b85c80850a9df3a4380a95e1a3af0949e47a3b
**CMSSW**: CMSSW_12_3_X_2021-12-21-2300/slc7_amd64_gcc10
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/36576/21442/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 43
 - DQMHistoTests: Total histograms compared: 3461692
 - DQMHistoTests: Total failures: 0
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3461670
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 42 files compared)
 - Checked 181 log files, 42 edm output root files, 43 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_3_X_2021-12-21-2300+2ab928/47636/triggerResults): no differences found

slava77 commented on RecoBTag/PerformanceDB/test/TestBtagPayloads.cc:
```diff
@@ -36,19 +36,94 @@
+  for (auto const& n : measureName_) {
+    if (lastName != n) {
+      lastName = n;
+      lastToken = esConsumes(edm::ESInputTag("", n));
```
is there some cost from calling `esConsumes` multiple times and more importantly, is there a change in logic if it's called multiple times?
- If there is no harm, then this optimization to skip is ok, 
- if the logic changes, then this check seems rather weak (assuming that `measureName_` is sorted, as in the current hardcoded implementation) and some more thorough check is needed (perhaps the easiest is to make a sorted temporary)

Dr15Jones replied:
There is only a minor performance hit.
- needs more memory to store the consumes info
- each identical consumes will create an additional entry in the WaitingTaskHolder for the item which takes more memory and a small bit of time.

Mostly, this bugs me aesthetically.

slava77 commented on RecoBTag/PerformanceDB/test/TestBtagPayloads.cc:
```diff
@@ -98,48 +168,11 @@ void TestBtagPayloads::analyze(const edm::Event& iEvent, const edm::EventSetup&
 
   printf("\033[22;31m TESTING FOR TTBAR WP's \n\033[0m");
 
-  //Possible algorithms: TTBARWPBTAGCSVL,  TTBARWPBTAGCSVM,   TTBARWPBTAGCSVT
-  //                     TTBARWPBTAGJPL,   TTBARWPBTAGJPM,    TTBARWPBTAGJPT
-  //                                                          TTBARWPBTAGTCHPT
```
```suggestion
  //Possible algorithms: TTBARWPBTAGCSVL,  TTBARWPBTAGCSVM,   TTBARWPBTAGCSVT
  //                     TTBARWPBTAGJPL,   TTBARWPBTAGJPM,    TTBARWPBTAGJPT
  //                                                          TTBARWPBTAGTCHPT
```
this comment is lost in full; perhaps instead of recovering here, better copy this up above before filling the `measureName_`

Dr15Jones replied:
Good catch.

Dr15Jones committed:
Changes based on review comments

- Restored original comment
- Made reduced esConsumes call algorithm more robust.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-36576/27542

- This PR adds an extra 12KB to repository

cmsbuild commented:
Pull request #36576 was updated. @jpata, @cmsbuild, @clacaputo, @slava77 can you please check and sign again.

Dr15Jones commented:
please test

cmsbuild commented:
-1

**Failed Tests**:   Build  ClangBuild 
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-2ab928/21456/summary.html
**COMMIT**: db19db45bfba9ba2bfbb679a2adbb21682e9a5f4
**CMSSW**: CMSSW_12_3_X_2021-12-23-1100/slc7_amd64_gcc10
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/36576/21456/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Build

I found compilation error when building: 

<pre>>> Compiling edm plugin /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_12_3_X_2021-12-23-1100/src/RecoBTag/PerformanceDB/test/DumpBtagTable.cc
>> Compiling edm plugin /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_12_3_X_2021-12-23-1100/src/RecoBTag/PerformanceDB/test/TestBtagPayloads.cc
>> Compiling edm plugin /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_12_3_X_2021-12-23-1100/src/RecoBTag/PerformanceDB/test/TestPerformanceFW_ES_TFormula.cc
>> Compiling edm plugin /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_12_3_X_2021-12-23-1100/src/RecoBTag/PerformanceDB/test/TestPerformanceFW_ES.cc
/data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_12_3_X_2021-12-23-1100/src/RecoBTag/PerformanceDB/test/TestBtagPayloads.cc: In constructor 'TestBtagPayloads::TestBtagPayloads(const edm::ParameterSet&)':
/data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_12_3_X_2021-12-23-1100/src/RecoBTag/PerformanceDB/test/TestBtagPayloads.cc:122:19: error: 'seenNames' was not declared in this scope
  122 |     auto insert = seenNames.insert(n, {});
      |                   ^~~~~~~~~
gmake: *** [tmp/slc7_amd64_gcc10/src/RecoBTag/PerformanceDB/test/testRecoBTagTestPerformanceDBPlugins/TestBtagPayloads.cc.o] Error 1
>> Building  edm plugin tmp/slc7_amd64_gcc10/src/RecoBTag/PerformanceDB/test/testRecoBTagTestPerformanceDBPlugins/libtestRecoBTagTestPerformanceDBPlugins.so
c++: error: tmp/slc7_amd64_gcc10/src/RecoBTag/PerformanceDB/test/testRecoBTagTestPerformanceDBPlugins/TestBtagPayloads.cc.o: No such file or directory
</pre>

## Clang Build

I found compilation error while trying to compile with clang. Command used:
```
USER_CUDA_FLAGS='--expt-relaxed-constexpr' USER_CXXFLAGS='-Wno-register -fsyntax-only' scram build -k -j 32 COMPILER='llvm compile'

```


<pre>>> Local Products Rules ..... started
>> Local Products Rules ..... done
>> Creating project symlinks
>> Entering Package RecoBTag/PerformanceDB
>> Compile sequence completed for CMSSW CMSSW_12_3_X_2021-12-23-1100
gmake: *** [There are compilation/build errors. Please see the detail log above.] Error 1
+ eval scram build outputlog '&&' '(python3' /data/cmsbld/jenkins/workspace/ib-run-pr-tests/cms-bot/buildLogAnalyzer.py --ignoreWarning=Wdeprecated-declarations --logDir /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_12_3_X_2021-12-23-1100/tmp/slc7_amd64_gcc10/cache/log/src '||' 'true)'
++ scram build outputlog
>> Entering Package RecoBTag/PerformanceDB
>> Compiling  /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_12_3_X_2021-12-23-1100/src/RecoBTag/PerformanceDB/bin/TestPerformanceFWLite_ES.cc
>> Compiling  /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_12_3_X_2021-12-23-1100/src/RecoBTag/PerformanceDB/bin/getbtagPerformance.cc
</pre>

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-36576/27543

- This PR adds an extra 12KB to repository

cmsbuild commented:
Pull request #36576 was updated. @jpata, @cmsbuild, @clacaputo, @slava77 can you please check and sign again.

Dr15Jones commented:
please test

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-2ab928/21457/summary.html
**COMMIT**: 48a4603fa7cb9b02eb896c00afd93f3527c1b003
**CMSSW**: CMSSW_12_3_X_2021-12-23-1100/slc7_amd64_gcc10
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/36576/21457/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 43
 - DQMHistoTests: Total histograms compared: 3461692
 - DQMHistoTests: Total failures: 0
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3461670
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 42 files compared)
 - Checked 181 log files, 42 edm output root files, 43 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_3_X_2021-12-23-1100+2ab928/47645/triggerResults): no differences found

slava77 commented:
+reconstruction

for https://github.com/cms-sw/cmssw/pull/36576	48a4603
- code changes are technical, in line with the PR description and the follow up review. IIUC, most of the code is not exercised in the PR tests.
- jenkins tests pass

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @perrotta, @dpiparo, @qliphy (and backports should be raised in the release meeting by the corresponding L2)

qliphy commented:
+1
